crates/core/tedge_actors/src/builders.rs

```diff
@@ -169,23 +169,30 @@ pub trait MessageSink<M: Message> {
         MessageMapper: Fn(N) -> MS,
         MessageMapper: 'static + Send + Sync,
     {
-        let sender = MappingSender::new(self.get_sender(), cast);
-        source.register_peer(config, sender.into())
+        let sender: DynSender<N> = MappingSender::new(self.get_sender(), cast).into();
+        source.connect_sink(config, &sender)
+    }
+}
+
+/// A [DynSender] can be used as a [MessageSink],
+/// provided the messages expected by the sender can be built from those sent to the sink.
+impl<N: Message, M: Message + From<N>> MessageSink<N> for DynSender<M> {
+    fn get_sender(&self) -> DynSender<N> {
+        self.sender_clone()
     }
 }
 
 /// The [Builder] of an [Actor](crate::Actor) must implement this trait
 /// for every message type that actor can send to its peers.
 ///
-/// To receive messages from a `MessageSource<M, C>`, the peer must be a `MessageSink<M>`.
+/// To receive messages from a `MessageSource<M, C>`, the peer must be a `MessageSink<From<M>`.
 pub trait MessageSource<M: Message, Config> {
-    /// The message will be sent to the peer using the provided `sender`
-    fn register_peer(&mut self, config: Config, sender: DynSender<M>);
-
-    /// Connect a peer actor that will consume the message produced by this actor
-    fn add_sink(&mut self, config: Config, peer: &impl MessageSink<M>) {
-        self.register_peer(config, peer.get_sender());
-    }
+    /// Connect a peer actor that will consume the messages produced by this actor
+    ///
+    /// The messages will be sent to the peer using its sender, the `peer.get_sender()`.
+    /// A peer can subscribe to a subset of the messages produced by this source.
+    /// This subset of messages expected by the peer is defined by the `config` parameter.
+    fn connect_sink(&mut self, config: Config, peer: &impl MessageSink<M>);
 }
```

Comment:
Now a `MessageSource` only has to define how to connect a `MessageSink`, so messages matching the subscription can be sent to the sink via its sender.